### ISSUE TITLE
question on Mask R CNN COCO dataset to Masks

### ISSUE THREAD

**Issue by EMCP** [akTwelve/tutorials#16]
Hey Adam, 

Great to reread these fantastic tutorials..  I'd been in other things but coming back around to working with PyTorch.  One question regarding your creation of the bit mask 

https://github.com/akTwelve/tutorials/blob/master/image_composition/BasicImageComposition.ipynb

Did you end up creating a channel in the mask image for each object id like some I have seen?  Or do you create the bit-mask in one layer.. and number each pixel the id of the object overwhich it's covering?  I guess both can work but.. want to conform as best as I can to norms.

thanks!

**Comment by akTwelve** [comment]
The second option is better. One mask per channel limits you to only 3 objects.
________________________________
From: Erik ***@***.***>
Sent: Monday, June 14, 2021 1:24:52 AM
To: akTwelve/tutorials ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [akTwelve/tutorials] question on Mask R CNN COCO dataset to Masks (#16)


Hey Adam,

Great to reread these fantastic tutorials.. I'd been in other things but coming back around to working with PyTorch. One question regarding your creation of the bit mask

https://github.com/akTwelve/tutorials/blob/master/image_composition/BasicImageComposition.ipynb<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FakTwelve%2Ftutorials%2Fblob%2Fmaster%2Fimage_composition%2FBasicImageComposition.ipynb&data=04%7C01%7C%7C979d6502839c4e5618da08d92efd1ee8%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637592486942080582%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=Vg8EU7%2BogpMhJNcqGp7%2Fnzlt8KMqCxkJequpoV7DuSQ%3D&reserved=0>

Did you end up creating a channel in the mask image for each object id like some I have seen? Or do you create the bit-mask in one layer.. and number each pixel the id of the object overwhich it's covering? I guess both can work but.. want to conform as best as I can to norms.

thanks!

**Comment by EMCP** [comment]
Sounds like a plan... I'm in pytorch these days and I will add the objects to a tensor like this example tutorial then

https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html

```
import os
import numpy as np
import torch
from PIL import Image


class PennFudanDataset(object):
    def __init__(self, root, transforms):
        self.root = root
        self.transforms = transforms
        # load all image files, sorting them to
        # ensure that they are aligned
        self.imgs = list(sorted(os.listdir(os.path.join(root, "PNGImages"))))
        self.masks = list(sorted(os.listdir(os.path.join(root, "PedMasks"))))

    def __getitem__(self, idx):
        # load images and masks
        img_path = os.path.join(self.root, "PNGImages", self.imgs[idx])
        mask_path = os.path.join(self.root, "PedMasks", self.masks[idx])
        img = Image.open(img_path).convert("RGB")
        # note that we haven't converted the mask to RGB,
        # because each color corresponds to a different instance
        # with 0 being background
        mask = Image.open(mask_path)
        # convert the PIL Image into a numpy array
        mask = np.array(mask)
        # instances are encoded as different colors
        obj_ids = np.unique(mask)
        # first id is the background, so remove it
        obj_ids = obj_ids[1:]

        # split the color-encoded mask into a set
        # of binary masks
        masks = mask == obj_ids[:, None, None]

        # get bounding box coordinates for each mask
        num_objs = len(obj_ids)
        boxes = []
        for i in range(num_objs):
            pos = np.where(masks[i])
            xmin = np.min(pos[1])
            xmax = np.max(pos[1])
            ymin = np.min(pos[0])
            ymax = np.max(pos[0])
            boxes.append([xmin, ymin, xmax, ymax])

        # convert everything into a torch.Tensor
        boxes = torch.as_tensor(boxes, dtype=torch.float32)
        # there is only one class
        labels = torch.ones((num_objs,), dtype=torch.int64)
        masks = torch.as_tensor(masks, dtype=torch.uint8)

        image_id = torch.tensor([idx])
        area = (boxes[:, 3] - boxes[:, 1]) * (boxes[:, 2] - boxes[:, 0])
        # suppose all instances are not crowd
        iscrowd = torch.zeros((num_objs,), dtype=torch.int64)

        target = {}
        target["boxes"] = boxes
        target["labels"] = labels
        target["masks"] = masks
        target["image_id"] = image_id
        target["area"] = area
        target["iscrowd"] = iscrowd

        if self.transforms is not None:
            img, target = self.transforms(img, target)

        return img, target

    def __len__(self):
        return len(self.imgs)
```
